### DISCUSSION TITLE
revalidating ssr page caches with renderToHTML failed after nextjs v12

### DISCUSSION THREAD

**Discussion by khanbaba** [vercel/next.js#35153] [Help]
### Run `next info` (available from version 12.0.8 and up)

    Operating System:
      Platform: linux
      Arch: x64
      Version: #29~20.04.1-Ubuntu SMP Fri Jan 14 00:32:30 UTC 2022
    Binaries:
      Node: 14.18.2
      npm: 6.14.15
      Yarn: 1.22.15
      pnpm: N/A
    Relevant packages:
      next: 12.0.9
      react: 17.0.2
      react-dom: 17.0.2

### What version of Next.js are you using?

12.0.9

### What version of Node.js are you using?

14.18.2

### What browser are you using?

Chrome

### What operating system are you using?

Ubuntu 20

### How are you deploying your application?

custom server with express.js

### Describe the Bug

We cached SSR pages in a redis database. after each hit on cache we want to revalidate the cache value in redis.
before migrating to next v12 every thing was ok. see the following code for revalidating cache:

``` js
async function revalidateCache(key, req) {
  try {
    const res = {}
    const value = await app.renderToHTML(
      { ...req, is_revalidate: true },
      res,
      '/browse',
      { ...getParamsFromUrl(req.url), ...req.query }
    )
    memcached_client.set(key, value)
  } catch (error) {
    console.log('revalidation error: ', error)
  }
}
```
which caused the following error:
`TypeError: Cannot read property 'Symbol(NextRequestMeta)' of undefined
    at Object.getRequestMeta (/home/torob/Desktop/Repositories/nextjs/node_modules/next/dist/server/request-meta.js:12:21)
    at Object.renderToHTML (/home/torob/Desktop/Repositories/nextjs/node_modules/next/dist/server/render.js:342:153)
    at async doRender (/home/torob/Desktop/Repositories/nextjs/node_modules/next/dist/server/base-server.js:855:38)
    at async /home/torob/Desktop/Repositories/nextjs/node_modules/next/dist/server/base-server.js:950:28
    at async /home/torob/Desktop/Repositories/nextjs/node_modules/next/dist/server/response-cache.js:63:36`

``` js

after removing is_validate as follow:

async function revalidateCache(key, req) {
  try {
    const res = {}
    const value = await app.renderToHTML(
      req,
      res,
      '/browse',
      { ...getParamsFromUrl(req.url), ...req.query }
    )
    memcached_client.set(key, value)
  } catch (error) {
    console.log('revalidation error: ', error)
  }
}
```

error changed to this:
`TypeError: Cannot read property 'finished' of undefined
    at Object.isResSent (/home/torob/Desktop/Repositories/nextjs/node_modules/next/dist/shared/lib/utils.js:40:16)
    at Object.renderToHTML (/home/torob/Desktop/Repositories/nextjs/node_modules/next/dist/server/render.js:660:21)
    at async doRender (/home/torob/Desktop/Repositories/nextjs/node_modules/next/dist/server/base-server.js:855:38)
    at async /home/torob/Desktop/Repositories/nextjs/node_modules/next/dist/server/base-server.js:950:28
    at async /home/torob/Desktop/Repositories/nextjs/node_modules/next/dist/server/response-cache.js:63:36`

it seems that renderToHTML needs real req and res and after res.end we could not get html to update the cache.

### Expected Behavior

I need a api like renderToHTML to get html from nextjs.

### To Reproduce

create a simple express custom server and for a route (for example /a) send a fake response as follow:

`res.send('hello world')`

after that try to get real html from nextjs with renderToHTML or render.

**Comment by code-freakz** [comment]
Hi,
have you got this fixed?
I'm also facing the same issue now
